### PR TITLE
Merge `lib` submodule into `ops`

### LINKED ISSUE #2032: Deduplicate ops/lib code
### Is your feature request related to a specific problem?

Now that `fideslib` has been merged into `fides`, there is some suboptimal logic duplication going on from where `ops` had to patch over `fideslib` functionality for specific use-cases.

### Describe the solution you'd like

A thorough refactor should be done to consolidate this code wherever possible

### Additional context

There are some examples in [this PR](), but a few examples would be:
* `oauth.util.py` that exists in both `lib` and `ops`
* `deps.py` that exists in both `lib` and `ops`

### PR DESCRIPTION
**Pull request by ThomasLaPiana**
Closes #2032 

### Code Changes

* [x] move source files from `lib` into `ops`, mostly maintaining the structure
* [x] remove the `lib` submodule completely once the functionality has been copied over
* [x] clean up `ops`/`lib` oauth utils/functions

### Steps to Confirm

* [ ] automated checks pass
* [ ] logical code conversion/synthesis is done

### Pre-Merge Checklist

* [ ] All CI Pipelines Succeeded
* ~~Documentation~~:
  * [ ] documentation complete, [PR opened in fidesdocs](https://github.com/ethyca/fidesdocs/pulls)
  * [ ] documentation [issue created in fidesdocs](https://github.com/ethyca/fidesdocs/issues/new/choose)
* [x] Issue Requirements are Met
* [x] Relevant Follow-Up Issues Created ([Fidesplus follow-up](https://github.com/ethyca/fidesplus/issues/869))
* [x] Update `CHANGELOG.md`
* [ ] ~~For API changes, the [Postman collection](https://github.com/ethyca/fides/blob/main/docs/fides/docs/development/postman/Fides.postman_collection.json) has been updated~~

### Description Of Changes

Long-awaited change to merge the previously-separate `lib` code with the `ops` api code. This is an iteration on the path towards a total unification of `ctl` and `ops` server code as well

Given the size of the changes here as well as being a pure refactor change, I'm not expecting much in terms of a manual review here

### CODE REVIEW

**Comment by ThomasLaPiana** [comment]
@pattisdr I'm nearing the point of being ready for a review, but before I get there I wanted to make sure there wasn't anything super specific you had in mind for cleanup here? At the very least this will combine the code and do some minimal de-duplication

**Comment by pattisdr** [comment]
@ThomasLaPiana taking a look at this now, apologize for missing this notification last week

**Comment by ThomasLaPiana** [comment]
ok...seems like it should be good to go now!

File: src/fides/api/ops/api/v1/endpoints/privacy_request_endpoints.py [hunk 1]
```diff
@@ -761,10 +761,12 @@ def get_request_preview_queries(
                 )
             dataset_configs.append(dataset_config)
     try:
-        connection_configs: List[ConnectionConfig] = [
+        connection_configs = [
             ConnectionConfig.get(db=db, object_id=dataset.connection_config_id)
             for dataset in dataset_configs
         ]
+        connection_configs = [config for config in connection_configs if config]
+        assert connection_configs
```
Comment (pattisdr) [inline src/fides/api/ops/api/v1/endpoints/privacy_request_endpoints.py:769]: Why is this added here?

File: src/fides/api/ops/models/connectionconfig.py [hunk 2]
```diff
@@ -125,6 +127,10 @@ class ConnectionConfig(Base):
     saas_config = Column(
         MutableDict.as_mutable(JSONB), index=False, unique=False, nullable=True
     )
+
+    # This is imported here to avoid a circular dependency
+    from fides.api.ctl.sql_models import System  # type: ignore[attr-defined]
+
     system_id = Column(
         String, ForeignKey(System.id_field_path), nullable=True, index=True
```
Comment (pattisdr) [inline src/fides/api/ops/models/connectionconfig.py:135]: Alternatively, you can define the FK like this:

```
  system_id = Column(
        String, ForeignKey("ctl_systems.id"), nullable=True, index=True
    )
```

and 

```
 system = relationship("System")
```

instead of importing System like this, but what suddenly prompted the circular dependency?

File: src/fides/api/ops/models/messaging.py [hunk 3]
```diff
@@ -97,14 +96,12 @@ class MessagingConfig(Base):
...
-    def get_configuration(cls, db: Session, service_type: str) -> Base:
+    def get_configuration(cls, db: Session, service_type: str) -> OrmWrappedFidesBase:
         """
         Fetches the configured MessagingConfig record by service type. Once fetched this function validates that
         the MessagingConfig is configured with secrets.
         """
-        instance: Optional[Base] = cls.get_by(
-            db=db, field="service_type", value=service_type
-        )
+        instance = cls.get_by(db=db, field="service_type", value=service_type)
```
Comment (pattisdr) [inline src/fides/api/ops/models/messaging.py:104]: so base_class seems to define `declarative_base(cls=OrmWrappedFidesBase)` what is prompting this OrmWrappedFidesBase usage?

File: src/fides/api/ops/oauth/api/deps.py [hunk 4]
```diff
@@ -2,10 +2,10 @@
...
+from fides.api.ops.db.session import get_db_session
+from fides.api.ops.oauth.api.urn_registry import TOKEN, V1_URL_PREFIX
+from fides.api.ops.oauth.schemas.oauth import OAuth2ClientCredentialsBearer
 from fides.core.config import get_config
-from fides.lib.db.session import get_db_session
-from fides.lib.oauth.api.urn_registry import TOKEN, V1_URL_PREFIX
-from fides.lib.oauth.schemas.oauth import OAuth2ClientCredentialsBearer
 
 
 def get_db() -> Generator[Session, None, None]:
```
Comment (pattisdr) [inline src/fides/api/ops/oauth/api/deps.py:11]: I think this file should be removed in favor of src/fides/api/ops/api/deps.py. Nothing seems to be using the oauth2_scheme defined in this file, instead preferring the definition here: src/fides/api/ops/util/oauth_util.py, and get_db() is similarly defined in src/fides/api/ops/api/deps.py

File: src/fides/api/ops/oauth/privileges.py
Comment (pattisdr) [inline src/fides/api/ops/oauth/privileges.py:1]: This empty file could be removed.  Before we had "roles" we were exploring the similar idea of "privileges" and that work ended up getting deprioritized and then removed later

File: src/fides/api/ops/oauth/schemas/oauth.py
Comment (pattisdr) [inline src/fides/api/ops/oauth/schemas/oauth.py:1]: These schemas for ops API endpoints tend to go in src/fides/api/ops/schemas instead.  Since a lot of these schemas are used in the oauth endpoints which are also defined in the typical location, I wouldn't have this here as an exception

File: src/fides/api/ops/oauth/schemas/user.py
Comment (pattisdr) [inline src/fides/api/ops/oauth/schemas/user.py:1]: Can we similarly move these user schemas to src/fides/api/ops/schemas

File: src/fides/api/ops/service/connectors/fides_connector.py [hunk 5]
```diff
@@ -182,6 +182,6 @@ def filter_fides_connector_datasets(
     return {
         dataset.fides_key
         for connector_config in connector_configs
-        for dataset in connector_config.datasets
+        for dataset in connector_config.datasets  # type: ignore[attr-defined]
```
Comment (pattisdr) [inline src/fides/api/ops/service/connectors/fides_connector.py:185]: ahhh why are these connection config attributes suddenly causing problems

File: src/fides/api/ops/service/messaging/message_dispatch_service.py [hunk 6]
```diff
@@ -138,7 +138,7 @@ def dispatch_message(
         raise MessageDispatchException("No notification service type configured.")
 
     logger.info("Retrieving message config")
-    messaging_config: MessagingConfig = MessagingConfig.get_configuration(
+    messaging_config = MessagingConfig.get_configuration(
```
Comment (pattisdr) [inline src/fides/api/ops/service/messaging/message_dispatch_service.py:141]: Noting a type hint being removed likely to get around linting issues

File: src/fides/api/ops/service/messaging/messaging_crud_service.py [hunk 7]
```diff
@@ -13,9 +13,7 @@
 def update_messaging_config(
     db: Session, key: FidesKey, config: MessagingConfigRequest
 ) -> MessagingConfigResponse:
-    existing_config_with_key: MessagingConfig = MessagingConfig.get_by(
-        db=db, field="key", value=key
-    )
+    existing_config_with_key = MessagingConfig.get_by(db=db, field="key", value=key)
```
Comment (pattisdr) [inline src/fides/api/ops/service/messaging/messaging_crud_service.py:16]: 👀

File: src/fides/api/ops/util/oauth_util.py [hunk 8]
```diff
@@ -43,6 +42,19 @@
...
+
+
+def is_token_expired(issued_at: datetime | None, token_duration_min: int) -> bool:
+    """Returns True if the datetime is earlier than token_duration_min ago."""
+    if not issued_at:
+        return True
+
+    return (datetime.now() - issued_at).total_seconds() / 60.0 > token_duration_min
+
+
```
Comment (pattisdr) [inline src/fides/api/ops/util/oauth_util.py:57]: If we have a brand new oauth folder this whole file is something that could be potentially consolidated into that folder

**Review by pattisdr — CHANGES_REQUESTED** [review]
nice work @ThomasLaPiana really appreciate the consolidation here.  Main things that stood out were the ConnectionConfig/MessagingConfig type issues and some of endpoints/URL's/schemas still in separate files from there rest of the ops endpoint details.

**Comment by ThomasLaPiana** [comment]
@pattisdr thank you for the thorough review! Will go through and make changes accordingly 😄

File: src/fides/api/ops/service/messaging/message_dispatch_service.py (hunk 6 above)
Comment (ThomasLaPiana) [reply src/fides/api/ops/service/messaging/message_dispatch_service.py:141]: yep....I also have no idea why, but I had to add _a lot_ of mypy workarounds and that makes me concerned, but not concerned enough to spend hours figuring out why 😅

File: src/fides/api/ops/models/connectionconfig.py [hunk 9]
```diff
@@ -209,12 +215,12 @@ def update_test_status(
         self.last_test_succeeded = test_status == ConnectionTestStatus.succeeded
         self.save(db)
 
-    def delete(self, db: Session) -> Optional[Base]:
+    def delete(self, db: Session) -> Optional[OrmWrappedFidesBase]:
```
[the remark this answers is not in the document]
Comment (ThomasLaPiana) [reply src/fides/api/ops/models/connectionconfig.py:218]: `mypy` complaints

File: tests/conftest.py [hunk 10]
```diff
@@ -24,30 +24,30 @@
...
     OWNER,
     VIEWER,
     VIEWER_AND_APPROVER,
 )
+from fides.api.ops.schemas.messaging.messaging import MessagingServiceType
+
+# from fides.api.ops.tasks.scheduled.scheduler import scheduler
+from fides.api.ops.util.cache import get_cache
+
+# from fides.core import api
```
[the remark this answers is not in the document]
Comment (ThomasLaPiana) [reply tests/conftest.py:48]: these look like leftover artifacts, removing

File: src/fides/api/ops/service/connectors/fides_connector.py (hunk 5 above)
Comment (ThomasLaPiana) [reply src/fides/api/ops/service/connectors/fides_connector.py:185]: idk 😢 but it was a lot

File: src/fides/api/ops/models/connectionconfig.py (hunk 2 above)
Comment (ThomasLaPiana) [reply src/fides/api/ops/models/connectionconfig.py:135]: some way it was configured before, I don't remember now exactly why it happened

File: src/fides/api/ops/api/v1/endpoints/oauth_endpoints.py [hunk 11]
```diff
@@ -228,9 +228,10 @@ def oauth_callback(code: str, state: str, db: Session = Depends(get_db)) -> None
             detail="No authentication request found for the given state.",
         )
 
-    connection_config: ConnectionConfig = ConnectionConfig.get_by(
+    connection_config = ConnectionConfig.get_by(
         db, field="key", value=authentication_request.connection_key
     )
+    assert connection_config
```
[the remark this answers is not in the document]
Comment (ThomasLaPiana) [reply src/fides/api/ops/api/v1/endpoints/oauth_endpoints.py:231]: if I do that it will break so we'd need another solution

File: src/fides/api/ops/db/base_class.py [hunk 12]
```diff
@@ -32,3 +48,276 @@ def process_result_value(self, value: str, _: Any | None) -> dict[str, Any] | No
...
+    key = FidesKey.validate(data.get("key")) if data.get("key") else None
+    if key is None:
+        name = data.get("name")
+        if name is None:
+            raise KeyValidationError(f"{cls_name} requires a name.")
+        key = FidesKey.validate(to_snake_case(name))
+    return key
+
+
+class FidesBase:
```
[the remark this answers is not in the document]
Comment (ThomasLaPiana) [reply src/fides/api/ops/db/base_class.py:68]: Oh, that is odd. I'm going to leave it for now since the scope is already large enough here, but as I slowly refactor ctl into ops I'll try to revisit that and clean it up

**Comment by ThomasLaPiana** [comment]
@pattisdr I've gone through and made all of the consolidation/refactor changes you mentioned!

The only ones left are the `mypy` failure ones and I'm honestly not sure what to do there...I could revert some of those changes and then add ignores instead?

**Comment by pattisdr** [comment]
thanks very much @ThomasLaPiana I plan to review first thing tomorrow. I know this touches a lot of files and you're anxious to merge -

**Review by pattisdr — COMMENTED** [review]
@ThomasLaPiana Thanks very much especially for consolidating the user endpoints, that's probably what I hoped to get most out of this.

The cause of the type issues were bothering me so I dug into it and it looks like we were previously importing the MessagingConfig and ConnectionConfig models from a different directory in lib than the other models.  I've tried to continue to clean up type hints in this corresponding PR that is pointed at your branch plus some other minor items I saw:

https://github.com/ethyca/fides/pull/3221

File: src/fides/api/ctl/routes/generate.py [hunk 13]
```diff
@@ -16,7 +16,7 @@
 from fides.api.ctl.sql_models import sql_model_map  # type: ignore[attr-defined]
 from fides.api.ctl.utils.api_router import APIRouter
 from fides.api.ops.api.v1 import scope_registry
-from fides.api.ops.util.oauth_util import verify_oauth_client_prod
+from fides.api.ops.oauth.utils import verify_oauth_client_prod
```
[the remark this answers is not in the document]
Comment (ThomasLaPiana) [reply src/fides/api/ctl/routes/generate.py:19]: makes sense, I'll get on that as soon as this gets merged!

**Comment by ThomasLaPiana** [comment]
@pattisdr This is ready for a final review, thanks again so much for the work you did in the type hints! I got those merged in and everything is looking good. I've also opened and cross-linked a Fidesplus issue for fixing things over there

**Review by pattisdr — APPROVED** [review]
Looks good to me @ThomasLaPiana you'll need another changelog update (group "changed" together) after getting this up to date one more time. Thanks again for all your efforts here.

**Comment by ThomasLaPiana** [comment]
> Looks good to me @ThomasLaPiana you'll need another changelog update (group "changed" together) after getting this up to date one more time. Thanks again for all your efforts here.

Thank you so much for your contributions here! This is only the beginning of the decisive tech-debt payoff 🙂

**Comment by ThomasLaPiana** [comment]
failed tests are expected, merging
